kubernetes/istio.go

```diff
@@ -326,15 +329,29 @@ func DestinationRuleHasMTLSEnabled(destinationRule *networking_v1.DestinationRul
 // ClusterNameFromIstiod attempts to resolve the cluster info of the "home" cluster where kiali is running
 // by inspecting the istiod deployment. Assumes that the istiod deployment is in the same cluster as the kiali pod.
 func ClusterNameFromIstiod(conf *config.Config, k8s ClientInterface) (string, error) {
-	// The "cluster_id" is set in an environment variable of
-	// the "istiod" deployment. Let's try to fetch it.
-	istiodDeployments, err := k8s.GetDeployments(conf.ExternalServices.Istio.RootNamespace, metav1.ListOptions{LabelSelector: "app=istiod"})
+	// Loop through all namespaces to find istiod deployments. We don't perform a single query for all namespaces
+	// because that works only if Kiali has cluster-wide access.
+	namespaces, err := k8s.GetNamespaces("")
```

Comment:
_Any_ client call that passes `""` for the namespace will fail when Kiali doesn't have cluster-wide access. Here's roughly what I think you'll want to do:

```
if conf.AllNamespacesAccessible {
deployments = k8s.GetDeployments("")
} else {
for namespace in conf.Deployment.AccessibleNamespaces {
  deployments = append(deployments, k8s.GetDeployments(namespace)
  }
}
```
Looking again at `AccessibleNamespaces`, I don't think that works for multi-cluster but don't bother with changing it right now.

Reply:
I think the suggested approach probably does work for MC. The default DS should likely apply to the Kiali cluster.